teradata/datadog_checks/teradata/check.py

```diff
@@ -0,0 +1,186 @@
+                'logdata': self.config.auth_data,
+                'user': self.config.username,
+                'password': self.config.password,
+                'https_port': str(self.config.https_port),
+                'sslmode': self.config.ssl_mode.upper(),
+                'sslprotocol': self.config.ssl_protocol,
+            }
+        )
+
+        global_tags = [
+            'teradata_server:{}'.format(self.instance.get('server')),
+            'teradata_port:{}'.format(self.instance.get('port', 1025)),
+        ]
+        self._tags = list(self.config.tags)
+        self._tags.extend(global_tags)
+        self._query_manager.tags = self._tags
+
+        self._tables_filter = create_tables_filter(self)
+
+    def _execute_query_raw(self, query):
+        # type: (AnyStr) -> Iterable[Sequence]
+        with closing(self._connection.cursor()) as cursor:
+            query = query.format(self.config.database)
+            cursor.execute(query)
+            if cursor.rowcount < 1:
+                self._query_errors += 1
+                self.log.warning('Failed to fetch records from query: `%s`.', query)
+                return None
+            for row in cursor.fetchall():
+                query_name = re.search(r'(DBC.[^\s]+)', query).group(1)
+                try:
+                    yield self._queries_processor(row, query_name)
+                except Exception as e:
+                    self.log.debug(
+                        'Unable to process row returned from query "%s", skipping row %s. %s', query_name, row, e
+                    )
+                    yield row
+
+    def _executor_error_handler(self, error):
+        # type: (AnyStr) -> AnyStr
+        self._query_errors += 1
+        return error
+
+    @contextmanager
+    def connect(self):
+        # type: () -> Iterator[teradatasql.connection]
+        conn = None
+        if TERADATASQL_IMPORT_ERROR:
+            self.log.error(
+                'Teradata SQL Driver module is unavailable. Please double check your installation and refer to the '
+                'Datadog documentation for more information. %s',
+                TERADATASQL_IMPORT_ERROR,
+            )
+            raise TERADATASQL_IMPORT_ERROR
+        self.log.info('Connecting to Teradata database %s on server %s.', self.config.database, self.config.server)
+        try:
+            conn = teradatasql.connect(self._connect_params)
+            self.log.info('Connected to Teradata.')
+            yield conn
+        except Exception as e:
+            self.log.error('Unable to connect to Teradata. %s.', e)
+            raise e
+        finally:
+            if conn:
+                conn.close()
+
+    def submit_health_checks(self):
+        # type: () -> None
+        connect_status = ServiceCheck.OK
+        query_status = ServiceCheck.CRITICAL if self._query_errors else ServiceCheck.OK
+
+        self.service_check(SERVICE_CHECK_QUERY, query_status, tags=self._tags)
```

Comment:
`can_query` only emits if `can_connect` is OK. Otherwise the check doesn't query the database so I didn't think it made sense for `can_query` to emit anything.

Reply:
And yes, if there is an earlier exception, `self._query_errors` increments and `can_query` is emitted with a CRITICAL status